source/common/http/conn_manager_impl.cc

```diff
@@ -2087,18 +2092,28 @@ void ConnectionManagerImpl::ActiveStream::recreateStream(
   // FilterState that we inherit, we'll end up copying this every time even though we could get
   // away with just resetting it to the HCM filter_state_.
   if (filter_state->hasDataAtOrAboveLifeSpan(StreamInfo::FilterState::LifeSpan::Request)) {
-    (*connection_manager_.streams_.begin())->filter_manager_.streamInfo().filter_state_ =
+    new_active_stream.filter_manager_.streamInfo().filter_state_ =
         std::make_shared<StreamInfo::FilterStateImpl>(
             filter_state->parent(), StreamInfo::FilterState::LifeSpan::FilterChain);
   }
 
   // Make sure that relevant information makes it from the original stream info
   // to the new one. Generally this should consist of all downstream related
   // data, and not include upstream related data.
-  (*connection_manager_.streams_.begin())
-      ->filter_manager_.streamInfo()
-      .setFromForRecreateStream(filter_manager_.streamInfo());
+  new_active_stream.filter_manager_.streamInfo().setFromForRecreateStream(
```

Comment:
hm, so fundamentally the problem here is we have stream state in an object which does not persist across streams.  It's a larger change but I'd be inclined to move this state from state_ to a "should drain connection" or "should_close_iff_http11" option in stream info, because for one we'd keep all the items which need to persist across recreateStream in one place, and it conveniently also addresses https://github.com/envoyproxy/envoy/issues/26473
WDYT?

Reply:
Switch to streamInfo(). shouldDrainConnectionUponCompletion() instead.

BTW, this PR doesn't seem to address https://github.com/envoyproxy/envoy/issues/26473 though. I think this PR is a necessary step to address https://github.com/envoyproxy/envoy/issues/26473, but there should be more work to do.